*** linux/include/linux/ktime.h 2006-06-18 03:49:35.000000000 +0200
--- UFS922_Orig/ktime.h 2009-02-10 19:09:48.000000000 +0100
***************
*** 258,263 ****
--- 258,269 ----

  #endif

+ static inline s64 ktime_to_us(const ktime_t kt)
+ {
+         struct timeval tv = ktime_to_timeval(kt);
+         return (s64) tv.tv_sec * USEC_PER_SEC + tv.tv_usec;
+ }
+
  /*
   * The resolution of the clocks. The resolution value is returned in
   * the clock_getres() system call to give application programmers an
