*** b/include/mtd/mtd-abi.h	2007-10-09 22:31:38.000000000 +0200
--- a/include/mtd/mtd-abi.h	2010-02-16 16:48:17.000000000 +0100
***************
*** 7,12 ****
--- 7,17 ----
  #ifndef __MTD_ABI_H__
  #define __MTD_ABI_H__
  
+ #ifndef __KERNEL__ /* Urgh. The whole point of splitting this out into
+ 		    separate files was to avoid #ifdef __KERNEL__ */
+ #define __user
+ #endif
+ 
  struct erase_info_user {
  	uint32_t start;
  	uint32_t length;
