*** linux/drivers/i2c/busses/i2c-st40-pio.c	2009-02-26 19:02:53.000000000 +0100
--- UFS910/i2c-st40-pio.c	2009-02-26 19:01:18.000000000 +0100
***************
*** 73,83 ****
  
  #elif defined(CONFIG_CPU_SUBTYPE_STB7100)
  
! #if defined(CONFIG_SH_STB7100_REF) || defined(CONFIG_SH_STB5202_REF)
  
! #define NR_I2C_BUSSES 2
  static pio_address i2c_address[NR_I2C_BUSSES] = {
  	{2,0,2,1},
  	{4,0,4,1}
  };
  
--- 73,84 ----
  
  #elif defined(CONFIG_CPU_SUBTYPE_STB7100)
  
! #if defined(CONFIG_SH_STB7100_REF)
  
! #define NR_I2C_BUSSES 3
  static pio_address i2c_address[NR_I2C_BUSSES] = {
  	{2,0,2,1},
+ 	{3,0,3,1},
  	{4,0,4,1}
  };
  
***************
*** 92,97 ****
--- 93,106 ----
  
  #endif
  
+ #elif defined(CONFIG_CPU_SUBTYPE_STX7200)
+ 
+ #define NR_I2C_BUSSES 2
+ static pio_address i2c_address[NR_I2C_BUSSES] = {
+        {4,0,4,1},
+        {7,6,7,7},
+ };
+ 
  #else
  #error Need to configure the default I2C pins for this chip
  #endif
