*** linux-sh4/sound/stm/pcm_player.c.orig	2009-06-15 23:11:19.000000000 +0200
--- linux-sh4/sound/stm/pcm_player.c	2009-06-15 23:11:12.000000000 +0200
***************
*** 1002,1007 ****
--- 1002,1014 ----
  
  	/* TODO */
  
+ 	/* from bugzilla */
+ 	/* STx7100 has a problem with 16/16 bits FIFO organization,
+ 	 * so we disable the 16 bits samples capability... */
+ /*	if (pcm_player->ver <= ver__AUD_PCMOUT__90_1_3)*/
+ 		snd_stm_pcm_player_hw.formats &= ~SNDRV_PCM_FMTBIT_S16_LE;
+ 
+ 
  	/* Create ALSA lowlevel device */
  
  	result = snd_device_new(card, SNDRV_DEV_LOWLEVEL, pcm_player,
