*** linux-sh4-2.6.23.17_stm23_0119-orig/arch/sh/kernel/process.c	2009-06-06 20:17:15.000000000 +0200
--- linux-sh4/arch/sh/kernel/process.c	2009-07-17 20:22:49.000000000 +0200
***************
*** 106,115 ****
  
  void machine_restart(char *__unused)
  {
  	/* Use watchdog timer to trigger reset */
! 	local_irq_disable();
  	watchdog_trigger_immediate();
  	while (1) {};
  }
  
  void machine_halt(void)
--- 106,123 ----
  
  void machine_restart(char *__unused)
  {
+ /* Dagobert: Taken from stlinux22 because stlinux23
+  * handling does not work on ufs910.
+  */
+ 	/* SR.BL=1 and invoke address error to let CPU reset (manual reset) */
+ 	asm volatile ("ldc %0, sr\n\t"
+ 		      "mov.l @%1, %0"::"r" (0x10000000), "r"(0x80000001));
+ 
  	/* Use watchdog timer to trigger reset */
! /*	local_irq_disable();
  	watchdog_trigger_immediate();
  	while (1) {};
+ */
  }
  
  void machine_halt(void)
